starknet-accounts/src/account.rs

```diff
@@ -0,0 +1,22 @@
+use async_trait::async_trait;
+use starknet_core::types::{AddTransactionResult, BlockId, UnsignedFieldElement};
+use std::error::Error;
+
+#[async_trait]
+pub trait Account {
+    type GetNonceError: Error + Send;
```

Comment:
From a user perspective, I find it weird to have different error types per `IAccount` functions. What's the advantage compared to having just a single error enum? If OZ extends the API and adds a new external function, we would have to add an error type here right?

Reply:
There are a few options here off my head:

- Exactly what I'm doing here: passing in error types as generic parameters
- The trait itself enforces an error type, which in turn wraps a custom inner error type supplied through generic param. Essentially the same stuff as option 1
- Same as option 2 but wraps a `&'static str` instead

Out of the 3 options, only option 3 is "interportable". `serde_json` uses it AFAIK.

Since it's obviously impossible to create an umbrella error type that covers every possible future signer error scenario without resorting to static strings or magic integers (remember, adding enum variants is considered a source breaking change in Rust), I think it really comes down to whether it's expected that users would process the errors they get, or they'll just print to log and forget about it, since structured error types are much easier to handle than strings or magic codes.

So I made the design decision to use option 1. I also looked at `ethers-rs` (as always), and they seem to be doing the same for error types.

Reply:
Oh I think I misunderstood your comment lol. You're talking about per-funciton and not per-implementation.

Reply:
OK so for the per-function error type, I think it also comes down to whether it's expected that users will handle errors based on type. It's obviously impossible to get a signer-related error when calling `get_nonce`, but if we use a single error enum, users would have to either:

- unnecessarily list all possible error types in `match` arms and discard; or
- use a catch-all discard (a bit of code smell)

I'm not sure if this is the ideal design here. An alternative would be to have a single umbrella enum with each variant corresponding to one "type" of functions. But in that case, adding a new function to the trait WILL be a breaking change (since you're adding enum variants.

Reply:
Makes sense. The argument of breaking for less people when using separate enums is solid. I imagine in a single enum type scenario, people would be using a catch-all more which would lead to things breaking silently.

Reply:
Thanks for the discussion as well!

Reply:
Yeah the thing I really don't like about the umbrella enum pattern is that it forces people to write catch-all match arms. Of course there's no right or wrong here, just a matter of preference. I do see libraries choosing one over another all the time :)